### PR TITLE
Add Transform Action for mods

### PR DESCRIPTION
**Pull request by itanasi**
Add Unique that lets you Transform a unit to another Unit

Removed Conditional Stats Costs

TODO: AI evaluation to use action

image
image

### CODE REVIEW

**Comment by itanasi** [comment]
1. It's using the same checks as Upgrade, so yes
2. Also yes. Right now has an explicit newunit movement=0 as part of the action

**Comment by itanasi** [comment]
Adding the conditional to all unit actions was a bit more scope than I wanted to handle at the moment

**Comment by itanasi** [comment]
Done

File: android/assets/jsons/translations/template.properties [hunk 1]
```diff
@@ -946,6 +946,10 @@ Add in capital =
 Add to [comment] = 
 Upgrade to [unitType] ([goldCost] gold) = 
 Upgrade to [unitType]\n([goldCost] gold, [resources]) = 
+Transform to [unitType] = 
+Transform to [unitType] ([cost] [stat]) =
```
Comment (yairm210) [inline android/assets/jsons/translations/template.properties:950]: Remove extra translations, they'll be part of future conditional

File: android/assets/jsons/translations/template.properties [hunk 2]
```diff
@@ -1330,6 +1334,8 @@ Can upgrade from [unit] =
 Can upgrade from: = 
 Upgrades to [upgradedUnit] = 
 Obsolete with [obsoleteTech] = 
+Can Transform to [upgradedUnit] = 
+Can Transform to [upgradedUnit] for [cost] [stat] =
```
Comment (yairm210) [inline android/assets/jsons/translations/template.properties:1338]: Ditto

File: core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt [hunk 3]
```diff
@@ -472,6 +474,77 @@ object UnitActions {
...
+    ): ArrayList<UnitAction>? {
+        if (!unit.baseUnit().hasUnique(UniqueType.CanTransform)) return null // can't upgrade to anything
+        val unitTile = unit.getTile()
+        val civInfo = unit.civInfo
+        val transformList = ArrayList<UnitAction>()
+        for (unique in unit.baseUnit().getMatchingUniques(UniqueType.CanTransform,
+            StateForConditionals(unit = unit, civInfo = civInfo, tile = unitTile))) {
+            val upgradedUnit = civInfo.getEquivalentUnit(unique.params[0])
+            val statCostOfUpgrade = 0
+            val costStat = Stat.valueOf("Gold")
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt:498]: Remove

File: core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt [hunk 4]
```diff
@@ -472,6 +474,77 @@ object UnitActions {
...
+            val resourceRequirementsDelta = Counter<String>()
+            for ((resource, amount) in unit.baseUnit().getResourceRequirements())
+                resourceRequirementsDelta.add(resource, -amount)
+            for ((resource, amount) in upgradedUnit.getResourceRequirements())
+                resourceRequirementsDelta.add(resource, amount)
+            val newResourceRequirementsString = resourceRequirementsDelta.entries
+                .filter { it.value > 0 }
+                .joinToString { "${it.value} {${it.key}}".tr() }
+
+            val title = if (newResourceRequirementsString.isEmpty())
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt:513]: Remove stat stuff (should all be addative strings anyway)

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt [hunk 5]
```diff
@@ -1235,7 +1235,7 @@ class CivilizationInfo : IsPartOfGameInfoSerialization {
             Stat.Culture -> policies.storedCulture
             Stat.Science -> {
                 if (tech.currentTechnology() == null) 0
-                else tech.remainingScienceToTech(tech.currentTechnology()!!.name)
+                else tech.researchOfTech(tech.currentTechnology()!!.name)
```
Comment (itanasi) [inline core/src/com/unciv/logic/civilization/CivilizationInfo.kt:1238]: Pretty sure this is how it should be, but meed to double check nothing is dependent on this

File: core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt [hunk 6]
```diff
@@ -472,6 +474,71 @@ object UnitActions {
...
+                        unit.copyStatisticsTo(resurrectedUnit)
+                    } else { // Managed to upgrade
+                        unit.copyStatisticsTo(newUnit)
+                        newUnit.currentMovement = 0f
+                    }
+                }.takeIf {
+                    (unit.currentMovement > 0
+                            && !unit.isEmbarked()
+                            && unit.canUpgrade(unitToUpgradeTo = upgradedUnit)
+                            )
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt:535]:
```suggestion
                    unit.currentMovement > 0
                            && !unit.isEmbarked()
                            && unit.canUpgrade(unitToUpgradeTo = upgradedUnit)
```

**Review by yairm210 — APPROVED** [review]
Looks good!

**Pull request merged**